File: scraper/core/similarity_check.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
===============================================================================
  TravaillerEnCi — scraper/core/similarity_check.py
  Contrôle anti-duplication (PARTIE 2 — §2.6 Contraintes de contenu)

  Objectif : garantir que la description Markdown réécrite par Gemini est
  bien UNIQUE et reformulée, jamais un copier-coller de la source officielle
  (risque de pénalité Google pour contenu dupliqué).

  Méthode :
    1. Normalisation : minuscules, accents supprimés, ponctuation/Markdown
       retirés, espaces uniformisés.
    2. Deux mesures complémentaires, on conserve le MAX :
       • ratio de séquence (difflib.SequenceMatcher sur les tokens) — détecte
         les copier-coller et reformulations trop fidèles ;
       • indice de Jaccard sur les n-grams de 4 tokens — robuste aux
         changements d'ordre des phrases.

  Seuil par défaut : 30 % (SIMILARITY_THRESHOLD). Au-dessus → la fiche est
  marquée `confidence='low'` et signalée à la modération manuelle pour
  réécriture (jamais publiée automatiquement : tout passe par /admin/exams).

  Usage :  from scraper.core.similarity_check import text_similarity
===============================================================================
"""

from __future__ import annotations

import re
import unicodedata
from difflib import SequenceMatcher

# Seuil de similarité au-delà duquel une fiche doit être réécrite (30 %).
SIMILARITY_THRESHOLD = 0.30

# Caractères "bruit" (Markdown, ponctuation, séparateurs).
_NOISE_RE = re.compile(r"[*#_`>\[\](){}|!?.,;:'\"«»<>/\\\n\r\t—–…·•]+")


def normalize(text: str) -> str:
    """Normalise un texte pour comparaison : minuscules, sans accents ni bruit."""
    if not text:
        return ""
    # Décomposition Unicode + suppression des accents.
    decomposed = unicodedata.normalize("NFD", text)
    no_accents = "".join(c for c in decomposed if not unicodedata.combining(c))
    cleaned = _NOISE_RE.sub(" ", no_accents)
    return re.sub(r"\s+", " ", cleaned).strip().lower()


def _ngrams(tokens: list[str], n: int = 4) -> set[str]:
    """n-grams de tokens (défaut 4, pour être insensible à l'ordre local)."""
    if len(tokens) < n:
        return {" ".join(tokens)}
    return {" ".join(tokens[i : i + n]) for i in range(len(tokens) - n + 1)}


def text_similarity(source: str, rewritten: str) -> float:
    """Similarité 0..1 entre le texte source et la réécriture (max des mesures).

    ~1.0  → copie quasi intégrale (à réécrire impérativement)
    >0.30 → au-delà du seuil : signaler à la modération
    <0.30 → reformulation satisfaisante

    Métriques combinées (max) : ratio de séquence LCS, Jaccard symétrique sur
    n-grams, et couverture (n-grams de la réécriture présents dans la source).
    """
    tokens_a = normalize(source).split()
    tokens_b = normalize(rewritten).split()
    if not tokens_a or not tokens_b:
        return 0.0

    # Mesure 1 : ratio de séquence (copier-coller / reformulation légère).
    sequence_ratio = SequenceMatcher(None, tokens_a, tokens_b).ratio()

    # Mesure 2 : Jaccard sur n-grams de 4 tokens (ordre des phrases modifié).
    source_grams = _ngrams(tokens_a)
    rewrite_grams = _ngrams(tokens_b)
    union = source_grams | rewrite_grams
    jaccard = len(source_grams & rewrite_grams) / len(union) if union else 0.0

    # Mesure 3 : COUVERTURE (directionnelle) — part des n-grams de la
    # RÉÉCRITURE présents dans la source. Pertinente quand la source (page
    # complète) est bien plus longue que la description : une copie partielle
    # reste détectée.
    coverage = (
        len(source_grams & rewrite_grams) / len(rewrite_grams)
        if rewrite_grams
        else 0.0
    )

    return max(sequence_ratio, jaccard, coverage)


def needs_rewrite(source: str, rewritten: str, threshold: float = SIMILARITY_THRESHOLD) -> bool:
    """True si la réécriture est trop proche de la source (copie)."""
    return text_similarity(source, rewritten) > threshold

File: src/lib/similarityCheck.ts
/**
 *  TravaillerEnCi — src/lib/similarityCheck.ts
 *  Contrôle anti-duplication (PARTIE 2 — §2.6) — miroir TypeScript de
 *  scraper/core/similarity_check.py (utilisé côté admin /admin/exams).
 *
 *  Le seuil est partagé : une réécriture dont la similarité avec la source
 *  dépasse 30 % doit être retravaillée avant publication.
 */
export const SIMILARITY_THRESHOLD = 0.3;

// Caractères « bruit » (Markdown, ponctuation, séparateurs).
const NOISE_RE =
  /[*#_`>\[\](){}|!?.,;:'"«»<>/\\\n\r\t—–…·•]+/g;

export function normalizeText(text: string): string {
  if (!text) return '';
  return text
    .normalize('NFD')
    .replace(/[\u0300-\u036f]/g, '') // suppression des accents
    .replace(NOISE_RE, ' ')
    .replace(/\s+/g, ' ')
    .trim()
    .toLowerCase();
}

function tokens(text: string): string[] {
  return normalizeText(text).split(/\s+/).filter(Boolean);
}

/** n-grams de tokens (défaut 4) — insensible à l'ordre local des phrases. */
function ngrams(list: string[], n = 4): Set<string> {
  const out = new Set<string>();
  if (list.length < n) {
    if (list.length > 0) out.add(list.join(' '));
    return out;
  }
  for (let i = 0; i <= list.length - n; i++) {
    out.add(list.slice(i, i + n).join(' '));
  }
  return out;
}

/**
 * Ratio de plus longue sous-séquence commune (2·LCS/(|a|+|b|)) sur les tokens
 * — détecte les copier-coller et reformulations trop fidèles. DP itératif à
 * deux lignes (mémoire bornée) ; les listes sont tronquées à 2000 tokens.
 */
function lcsRatio(a: string[], b: string[]): number {
  const A = a.slice(0, 2000);
  const B = b.slice(0, 2000);
  if (A.length === 0 || B.length === 0) return 0;
  let prev = new Array<number>(B.length + 1).fill(0);
  for (let i = 1; i <= A.length; i++) {
    const curr = new Array<number>(B.length + 1).fill(0);
    for (let j = 1; j <= B.length; j++) {
      curr[j] = A[i - 1] === B[j - 1] ? prev[j - 1] + 1 : Math.max(prev[j], curr[j - 1]);
    }
    prev = curr;
  }
  return (2 * prev[B.length]) / (A.length + B.length);
}

/** Similarité 0..1 entre le texte source et la réécriture (max des mesures). */
export function textSimilarity(source: string, rewritten: string): number {
  const a = tokens(source);
  const b = tokens(rewritten);
  if (a.length === 0 || b.length === 0) return 0;

  const seq = lcsRatio(a, b);

  const sourceGrams = ngrams(a);
  const descGrams = ngrams(b);
  let inter = 0;
  for (const g of descGrams) if (sourceGrams.has(g)) inter++;
  const union = new Set([...descGrams, ...sourceGrams]);
  const jaccard = union.size === 0 ? 0 : inter / union.size;
  // Couverture (DIRECTIONNELLE) : part des n-grams de la DESCRIPTION présents
  // dans la source — pertinente quand la source (page complète) est bien plus
  // longue que la description : une copie partielle reste détectée.
  const coverage = descGrams.size === 0 ? 0 : inter / descGrams.size;

  return Math.max(seq, jaccard, coverage);
}

/** True si la réécriture est trop proche de la source (> seuil). */
export function needsRewrite(
  source: string,
  rewritten: string,
  threshold: number = SIMILARITY_THRESHOLD,
): boolean {
  return textSimilarity(source, rewritten) > threshold;
}
